src/main/scala/mesosphere/util/ZookeeperDirectState.scala

```diff
@@ -0,0 +1,186 @@
+  }
+
+  /**
+    * Unmarshall an Entry from Zookeeper and return it as an eventual Variable
+    */
+  private def handleFetch(name: String): Future[Variable] = Future {
+
+    if (!exists(name)) {
+      new LocalVariable(name)
+    }
+    else {
+      val data: Array[Byte] = client.getData.forPath(fullPath(name))
+      data.size match {
+        case 0 =>
+          new LocalVariable(name)
+        case _ =>
+          val entry = Protos.Entry.parseFrom(data)
+          new LocalVariable(name, entry.getValue.toByteArray, Some(entry.getUuid.toStringUtf8))
+      }
+    }
+  }
+
+  /**
+    * Marshall a variable to an Entry and store it in Zookeeper
+    */
+  private def handleStore(lv: LocalVariable): Future[Variable] = Future {
+
+    val persistedVariable = fetch(lv.name).get.asInstanceOf[LocalVariable]
+    val inputMatchesPersistedVersion = persistedVariable.storedUuid == lv.storedUuid
+
+    if (inputMatchesPersistedVersion) {
+
+      val entryData = mesosphere.marathon.Protos.Entry.newBuilder()
+        .setName(lv.name)
+        .setUuid(ByteString.copyFromUtf8(UUID.randomUUID().toString))
+        .setValue(ByteString.copyFrom(lv.value()))
+        .build()
+        .toByteArray
+
+      mkDir(Some(lv.name))
+      client.setData().forPath(fullPath(lv.name), entryData)
+      fetch(lv.name).get()
+    }
+    else {
+      log.warn(s"Attempt was made to persist ${lv.name} from obsolete version")
+      null
+    }
+
+  }
+
+  /**
+    * Remove the variable from our Zookeeper node. Returns false if deletion was not necessary.
+    */
+  private def handleExpunge(lv: LocalVariable): Future[Boolean] = Future {
+
+    if (exists(lv.name)) {
+      client.delete().forPath(fullPath(lv.name))
+      true
+    }
+    else false
+  }
+
+  /**
+    * Return the children of a Zookeeper node
+    */
+  private def handleNames: Future[util.Iterator[String]] = {
+    Future { client.getChildren.forPath(zNode).iterator() }
+  }
+
+  private def fullPath(name: String) = s"$zNode/$name"
+
+  private def mkDir(name: Option[String] = None, zkClient: CuratorFramework = client): Unit = {
+    val path = name map fullPath getOrElse zNode
+    ZKPaths.mkdirs(zkClient.getZookeeperClient.getZooKeeper, path)
+  }
+
+  private def exists(name: String): Boolean = {
+    client.checkExists().forPath(fullPath(name)) != null
+  }
+
+  /**
+    * Creates our Zookeeper client
+    */
+  private def createZookeeperClient(): Future[CuratorFramework] = {
+
+    val RETRY_POLICY = new ExponentialBackoffRetry(1000, 4)
+
+    val client = CuratorFrameworkFactory.newClient(
+      serverList,
+      sessionTimeout.toMillis.toInt,
+      connectionTimeout.toMillis.toInt,
+      RETRY_POLICY
+    )
+
+    client.start()
+
+    // Make sure we have a successful connection before returning the client
+    Future {
+      val connected = client.getZookeeperClient.blockUntilConnectedOrTimedOut()
```

Comment:
Wouldn't it be easier to create the `CuratorFramework` outside of this class and pass it in as parameter? Doing it in here makes the whole class more complex and blocks 2 threads.